File: android-capture/facescapture/src/main/java/ai/unlikeother/facescapture/FaceAnalyzer.kt
package ai.unlikeother.facescapture

import android.graphics.Bitmap
import android.graphics.Rect
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import kotlinx.coroutines.tasks.await
import kotlin.math.abs

internal class FaceAnalyzer {
    private val detector = FaceDetection.getClient(
        FaceDetectorOptions.Builder()
            .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
            .build()
    )

    suspend fun analyze(bitmap: Bitmap, targetPose: CapturePose, previousRect: FaceRect?): CaptureState {
        val faces = detector.process(InputImage.fromBitmap(bitmap, 0)).await()
        if (faces.isEmpty()) {
            return CaptureState(targetPose = targetPose, blockReason = CaptureBlockReason.NO_FACE)
        }
        if (faces.size > 1) {
            return CaptureState(targetPose = targetPose, blockReason = CaptureBlockReason.MULTIPLE_FACES)
        }

        val face = faces.first()
        val rect = face.boundingBox.toNormalized(bitmap.width, bitmap.height)
        val insideGuide = isInsideGuide(rect)
        val lightingOk = lightingScore(bitmap) >= 0.34f
        val sharpnessOk = sharpnessScore(bitmap) >= 0.018f
        val detectedPose = detectPose(face)
        val stable = isStable(rect, previousRect, detectedPose, targetPose)

        val blockReason = when {
            !lightingOk -> CaptureBlockReason.BAD_LIGHTING
            !sharpnessOk -> CaptureBlockReason.TOO_BLURRY
            !insideGuide -> CaptureBlockReason.OUT_OF_FRAME
            detectedPose != targetPose -> CaptureBlockReason.WRONG_POSE
            !stable -> CaptureBlockReason.HOLD_STILL
            else -> CaptureBlockReason.NONE
        }

        return CaptureState(
            targetPose = targetPose,
            detectedPose = detectedPose,
            faceRect = rect,
            faceInsideGuide = insideGuide,
            lightingOk = lightingOk,
            sharpnessOk = sharpnessOk,
            stable = stable,
            canCapture = blockReason == CaptureBlockReason.NONE,
            blockReason = blockReason
        )
    }

    private fun detectPose(face: Face): CapturePose {
        val yaw = face.headEulerAngleY
        val pitch = face.headEulerAngleX

        return when {
            yaw <= -8f && pitch <= -4f -> CapturePose.LEFT_TOP
            yaw >= 8f && pitch <= -4f -> CapturePose.TOP_RIGHT
            yaw >= 8f && pitch >= 6f -> CapturePose.BOTTOM_RIGHT
            yaw <= -8f && pitch >= 6f -> CapturePose.BOTTOM_LEFT
            abs(yaw) < 8f && pitch <= -4f -> CapturePose.TOP
            else -> CapturePose.STRAIGHT
        }
    }

    private fun isInsideGuide(rect: FaceRect): Boolean {
        val centerX = rect.x + rect.width / 2f
        val centerY = rect.y + rect.height / 2f
        return rect.width in 0.22f..0.58f &&
            rect.height in 0.22f..0.70f &&
            abs(centerX - 0.5f) <= 0.18f &&
            abs(centerY - 0.5f) <= 0.18f
    }

    private fun isStable(current: FaceRect,
                         previous: FaceRect?,
                         detectedPose: CapturePose,
                         targetPose: CapturePose): Boolean {
        if (previous == null || detectedPose != targetPose) return false
        val dx = abs((current.x + current.width / 2f) - (previous.x + previous.width / 2f))
        val dy = abs((current.y + current.height / 2f) - (previous.y + previous.height / 2f))
        val ds = abs(current.width - previous.width)
        return dx < 0.03f && dy < 0.03f && ds < 0.04f
    }

    private fun lightingScore(bitmap: Bitmap): Float {
        var total = 0f
        var count = 0
        val step = 16
        for (y in 0 until bitmap.height step step) {
            for (x in 0 until bitmap.width step step) {
                val pixel = bitmap.getPixel(x, y)
                val r = (pixel shr 16 and 0xff).toFloat()
                val g = (pixel shr 8 and 0xff).toFloat()
                val b = (pixel and 0xff).toFloat()
                total += (0.299f * r + 0.587f * g + 0.114f * b) / 255f
                count++
            }
        }
        return if (count == 0) 0f else total / count
    }

    private fun sharpnessScore(bitmap: Bitmap): Float {
        var total = 0f
        var count = 0
        val step = 20
        for (y in 0 until bitmap.height - step step step) {
            for (x in 0 until bitmap.width - step step step) {
                val p1 = bitmap.getPixel(x, y)
                val p2 = bitmap.getPixel(x + step, y)
                val p3 = bitmap.getPixel(x, y + step)
                val l1 = luma(p1)
                val l2 = luma(p2)
                val l3 = luma(p3)
                total += abs(l1 - l2) + abs(l1 - l3)
                count += 2
            }
        }
        return if (count == 0) 0f else (total / count) / 255f
    }

    private fun luma(pixel: Int): Float {
        val r = (pixel shr 16 and 0xff).toFloat()
        val g = (pixel shr 8 and 0xff).toFloat()
        val b = (pixel and 0xff).toFloat()
        return 0.299f * r + 0.587f * g + 0.114f * b
    }

    private fun Rect.toNormalized(width: Int, height: Int) = FaceRect(
        x = left.toFloat() / width,
        y = top.toFloat() / height,
        width = this.width().toFloat() / width,
        height = this.height().toFloat() / height,
    )
}

File: ios-capture/Sources/FacesCaptureKit/FaceAnalyzer.swift
import CoreGraphics
import CoreImage
import ImageIO
import Vision

@available(macOS 10.15, *)
struct FaceAnalysis {
    let state: CaptureState
    let landmarks: [CGPoint] // capture device normalized coordinates
}

@available(macOS 10.15, *)
final class FaceAnalyzer {
    private let request = VNDetectFaceLandmarksRequest()

    func analyze(buffer: CVPixelBuffer, targetPose: CapturePose, previousRect: FaceRect?) -> FaceAnalysis {
        let orientation = CGImagePropertyOrientation.leftMirrored
        let handler = VNImageRequestHandler(cvPixelBuffer: buffer, orientation: orientation, options: [:])

        do {
            try handler.perform([request])
        } catch {
            return FaceAnalysis(state: CaptureState(targetPose: targetPose, blockReason: .noFace), landmarks: [])
        }

        guard let faces = request.results, !faces.isEmpty else {
            return FaceAnalysis(state: CaptureState(targetPose: targetPose, blockReason: .noFace), landmarks: [])
        }
        guard faces.count == 1, let face = faces.first else {
            return FaceAnalysis(state: CaptureState(targetPose: targetPose, blockReason: .multipleFaces), landmarks: [])
        }

        let rect = FaceRect(
            x: face.boundingBox.minX,
            y: face.boundingBox.minY,
            width: face.boundingBox.width,
            height: face.boundingBox.height
        )
        let insideGuide = isInsideGuide(rect)
        let lightingOk = lightingScore(buffer) >= 0.25
        let sharpnessOk = sharpnessScore(buffer) >= 0.018
        let poseResult = detectPose(face)
        let detectedPose = poseResult.pose
        let stable = isStable(current: rect, previous: previousRect, detectedPose: detectedPose, targetPose: targetPose)

        let blockReason: CaptureBlockReason
        if !lightingOk {
            blockReason = .badLighting
        } else if !sharpnessOk {
            blockReason = .tooBlurry
        } else if !insideGuide {
            blockReason = .outOfFrame
        } else if detectedPose != targetPose {
            blockReason = .wrongPose
        } else if !stable {
            blockReason = .holdStill
        } else {
            blockReason = .none
        }

        let landmarkPoints = extractLandmarks(face)

        return FaceAnalysis(
            state: CaptureState(
                targetPose: targetPose,
                detectedPose: detectedPose,
                faceRect: rect,
                faceInsideGuide: insideGuide,
                lightingOk: lightingOk,
                sharpnessOk: sharpnessOk,
                stable: stable,
                canCapture: blockReason == .none,
                blockReason: blockReason,
                yaw: poseResult.yaw,
                verticalRatio: poseResult.verticalRatio
            ),
            landmarks: landmarkPoints
        )
    }

    private struct PoseResult {
        let pose: CapturePose
        let yaw: Double
        let verticalRatio: Double
    }

    private func detectPose(_ face: VNFaceObservation) -> PoseResult {
        let yaw = face.yaw?.doubleValue ?? 0
        let landmarks = face.landmarks

        guard
            let leftEye = averagePoint(landmarks?.leftEye),
            let rightEye = averagePoint(landmarks?.rightEye),
            let nose = averagePoint(landmarks?.nose),
            let outerLips = averagePoint(landmarks?.outerLips)
        else {
            let fallback: CapturePose
            if yaw < -0.18 { fallback = .topRight }
            else if yaw > 0.18 { fallback = .leftTop }
            else { fallback = .straight }
            return PoseResult(pose: fallback, yaw: yaw, verticalRatio: 0)
        }

        let eyesMidX = (leftEye.x + rightEye.x) / 2
        let eyeY = (leftEye.y + rightEye.y) / 2
        let mouthY = outerLips.y
        let horizontal = nose.x - eyesMidX
        let verticalRatio = (mouthY - eyeY)

        let vertical: Int
        if verticalRatio < 0.33 {
            vertical = 1
        } else if verticalRatio > 0.38 {
            vertical = -1
        } else {
            vertical = 0
        }

        let horizontalBucket: Int
        if horizontal < -0.02 || yaw > 0.12 {
            horizontalBucket = -1
        } else if horizontal > 0.02 || yaw < -0.12 {
            horizontalBucket = 1
        } else {
            horizontalBucket = 0
        }

        let pose: CapturePose
        switch (horizontalBucket, vertical) {
        case (-1, 1): pose = .leftTop
        case (1, 1): pose = .topRight
        case (1, -1): pose = .bottomRight
        case (-1, -1): pose = .bottomLeft
        case (0, 1): pose = .top
        default: pose = .straight
        }
        return PoseResult(pose: pose, yaw: yaw, verticalRatio: verticalRatio)
    }

    private func averagePoint(_ region: VNFaceLandmarkRegion2D?) -> CGPoint? {
        guard let region, region.pointCount > 0 else { return nil }
        let points = region.normalizedPoints
        let sum = points.reduce(CGPoint.zero) { partial, point in
            CGPoint(x: partial.x + CGFloat(point.x), y: partial.y + CGFloat(point.y))
        }
        return CGPoint(x: sum.x / CGFloat(points.count), y: sum.y / CGFloat(points.count))
    }

    /// Extract key landmark points and convert from Vision .leftMirrored space to capture device coordinates.
    private func extractLandmarks(_ face: VNFaceObservation) -> [CGPoint] {
        guard let lm = face.landmarks else { return [] }
        let bbox = face.boundingBox
        var points: [CGPoint] = []

        func toDevice(_ faceRelative: CGPoint) -> CGPoint {
            let vx = bbox.origin.x + faceRelative.x * bbox.width
            let vy = bbox.origin.y + faceRelative.y * bbox.height
            return CGPoint(x: 1 - vy, y: vx)
        }

        if let p = averagePoint(lm.leftEye) { points.append(toDevice(p)) }
        if let p = averagePoint(lm.rightEye) { points.append(toDevice(p)) }
        if let p = averagePoint(lm.nose) { points.append(toDevice(p)) }

        if let outerLips = lm.outerLips, outerLips.pointCount >= 2 {
            let pts = outerLips.normalizedPoints
            var minPt = pts[0], maxPt = pts[0]
            for i in 1..<outerLips.pointCount {
                if pts[i].x < minPt.x { minPt = pts[i] }
                if pts[i].x > maxPt.x { maxPt = pts[i] }
            }
            points.append(toDevice(CGPoint(x: CGFloat(minPt.x), y: CGFloat(minPt.y))))
            points.append(toDevice(CGPoint(x: CGFloat(maxPt.x), y: CGFloat(maxPt.y))))
        }

        return points
    }

    private func isInsideGuide(_ rect: FaceRect) -> Bool {
        let centerX = rect.x + rect.width / 2
        let centerY = rect.y + rect.height / 2
        return rect.width >= 0.22 &&
            rect.width <= 0.58 &&
            rect.height >= 0.22 &&
            rect.height <= 0.70 &&
            abs(centerX - 0.5) <= 0.18 &&
            abs(centerY - 0.5) <= 0.18
    }

    private func isStable(current: FaceRect,
                          previous: FaceRect?,
                          detectedPose: CapturePose,
                          targetPose: CapturePose) -> Bool {
        guard let previous, detectedPose == targetPose else { return false }
        let dx = abs((current.x + current.width / 2) - (previous.x + previous.width / 2))
        let dy = abs((current.y + current.height / 2) - (previous.y + previous.height / 2))
        let ds = abs(current.width - previous.width)
        return dx < 0.03 && dy < 0.03 && ds < 0.04
    }

    private func lightingScore(_ buffer: CVPixelBuffer) -> Double {
        CVPixelBufferLockBaseAddress(buffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(buffer, .readOnly) }
        guard let base = CVPixelBufferGetBaseAddress(buffer) else { return 0 }
        let width = CVPixelBufferGetWidth(buffer)
        let height = CVPixelBufferGetHeight(buffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
        let pixels = base.assumingMemoryBound(to: UInt8.self)
        var total = 0.0
        var count = 0.0
        let step = 16
        for y in Swift.stride(from: 0, to: height, by: step) {
            for x in Swift.stride(from: 0, to: width, by: step) {
                let offset = y * bytesPerRow + x * 4
                let b = Double(pixels[offset])
                let g = Double(pixels[offset + 1])
                let r = Double(pixels[offset + 2])
                total += (0.114 * b + 0.587 * g + 0.299 * r) / 255.0
                count += 1
            }
        }
        return count == 0 ? 0 : total / count
    }

    private func sharpnessScore(_ buffer: CVPixelBuffer) -> Double {
        CVPixelBufferLockBaseAddress(buffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(buffer, .readOnly) }
        guard let base = CVPixelBufferGetBaseAddress(buffer) else { return 0 }
        let width = CVPixelBufferGetWidth(buffer)
        let height = CVPixelBufferGetHeight(buffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(buffer)
        let pixels = base.assumingMemoryBound(to: UInt8.self)
        let step = 20
        var total = 0.0
        var count = 0.0
        for y in Swift.stride(from: 0, to: max(0, height - step), by: step) {
            for x in Swift.stride(from: 0, to: max(0, width - step), by: step) {
                let o1 = y * bytesPerRow + x * 4
                let o2 = y * bytesPerRow + (x + step) * 4
                let o3 = (y + step) * bytesPerRow + x * 4
                let l1 = 0.114 * Double(pixels[o1]) + 0.587 * Double(pixels[o1 + 1]) + 0.299 * Double(pixels[o1 + 2])
                let l2 = 0.114 * Double(pixels[o2]) + 0.587 * Double(pixels[o2 + 1]) + 0.299 * Double(pixels[o2 + 2])
                let l3 = 0.114 * Double(pixels[o3]) + 0.587 * Double(pixels[o3 + 1]) + 0.299 * Double(pixels[o3 + 2])
                total += Swift.abs(l1 - l2) + Swift.abs(l1 - l3)
                count += 2
            }
        }
        return count == 0 ? 0 : (total / count) / 255.0
    }
}
